### ISSUE TITLE
The pressed state of the "play" and "pause" buttons are not provided in Msaa.

### ISSUE THREAD

**Issue by paulber19** [audacity/audacity#4991]
### Bug description

Since version 3.2.2, it is no longer possible to know the pressed state of the "play", "Pause" buttons using MSAA. Only the name of the button indicates that the button is pressed, but this name is a translatable string and varies according to the language chosen by the user.
Knowing the status of these buttons is the only way to know that playback is in progress.
Until version 3.2.1 this was possible.

### Steps to reproduce

Using NVDA:
- focus playback button and press it
- - hit "nvda+f1" and look at state items.

### Expected behavior

Pressed button state must be accible by MSAA

### Actual behavior

Pressed button state is not available

### Audacity Version

latest stable version (from audacityteam.org/download)

### Operating system

Windows

### Additional context

ras

**Comment by DavidBailes** [comment]
Hi Paul,
let me explain why these toggle buttons were changed.
Since the introduction of realtime effects, these toggle buttons are now also used as power toggle buttons in both the realtime effects pane, and in realtime effect dialogs, and it was important that they behaved for all the windows screen readers, Jaws, NVDA, and Narrator, something like how toggle buttons normally behave. The problems are that:
1. Narrator does not read the pressed state at all.
2. For Jaws and NVDA, the not pressed state is not explicitly read.
Because of this, I changed the toggle button so that the name of the button, its role and its state are included in the accessibility name.
Users can still read the state of the buttons, and the overall playback status is still provided in the status bar.
However, I can understand that there is now a problem robustly getting the state of of one of these buttons from a screen reader script.
One suggestion for solving this would for me to append a special character to the end of the accessibility name of the toggle button when it was in the pressed state. For example the bell character '\a' which is not read by screen readers. Then you could check the last character in the accessibility name to see whether it was pressed.
If this would be helpful, just let me know and I can put in a pull request for this. Or alternatively you may have some other suggestions for how to work around this problem.
David.

**Comment by paulber19** [comment]
Hello David.
I fully understand your intention and your proposal is valid.
But what I don't understand is why indicating the "Pressed" state in the 
button name excludes positioning this state in the button state as it 
was in version 3.0.
Today, I worked around the problem by counting the word count of the 
button name, but that's not a reliable solution because it depends on 
the translation of that name.
I think it would be better to continue indicating this state as is the 
case for the "record" and "Stop" buttons and if possible, to emit an 
event when changing this state so that the readers of screen informs the 
user.
Best regards.
Paul.

**Comment by DavidBailes** [comment]
Hi Paul,
I'm leaving aside you comments about events just for the moment.
The reason I omitted the pressed state is that I thought that the consequent repetition of the word pressed would be annoying. This repetition issue if the pressed state is present varies with screen reader. I'll use the power button in the realtime effects panel as an example.
For Narrator, there's no repetition, as the pressed state is not read.
If a uses tabs to the button in the pressed state, then both Jaws and NVDA read "power button pressed pressed".
If the Power button is the focus and in the not pressed state, then when a user presses Enter, Jaws reads "power button pressed pressed" In fact it reads this whole name twice for reasons unknown. However NVDA does not repeat the word pressed, and just reads "power button pressed".
What are you views about this repetition issue if the pressed state is set?
David.

**Comment by paulber19** [comment]
Hi David.

> For Narrator, there's no repetition, as the pressed state is not read.
Maybe it's because the role of the button doesn't specify that it's a 
toggle button
and that NVDA and Jaws only care about the state of the button without 
associating it with the role.

> If a uses tabs to the button in the pressed state, then both Jaws and 
> NVDA read "power button pressed pressed".
It seems normal to me. NVDA and Jaws announces the button name, followed 
by the role, followed by the state,

> If the Power button is the focus and in the not pressed state, then 
> when a user presses Enter, Jaws reads "power button pressed pressed" 
> In fact it reads this whole name twice for reasons unknown.
I don't have an explanation either and not having your test version, I 
can't investigate.
Can you provide me with a download link for this version?
Thanks.

Frankly, I think it would be better to go back to the behavior of 
version 3.2.1 and:
- to give the "toggleButton" role to the button,
- to emit the event which signals the change of state (stateChange) when 
the "Enter" key is pressed (this should be possible since today, it is 
done for the change of name).

Best regards.
Paul.

**Comment by DavidBailes** [comment]
Hi Paul,
>> For Narrator, there's no repetition, as the pressed state is not read.

> Maybe it's because the role of the button doesn't specify that it's a toggle button and that NVDA and Jaws only care about the state of the button without associating it with the role.

Unfortunately, using MSAA, there is no toggle button role available.

>> If a uses tabs to the button in the pressed state, then both Jaws and NVDA read "power button pressed pressed".

> It seems normal to me. NVDA and Jaws announces the button name, followed by the role, followed by the state,

But for Jaws and NVDA users, being read "power button pressed" rather than "power button pressed pressed" is a preferable user experience.

>> If the Power button is the focus and in the not pressed state, then when a user presses Enter, Jaws reads "power button pressed pressed" In fact it reads this whole name twice for reasons unknown.

> I don't have an explanation either and not having your test version, I can't investigate. Can you provide me with a download link for this version?

I've created a branch with the pressed state also being set. The builds for this branch are available on this page of my repository:
https://github.com/DavidBailes/audacity/actions/runs/5865462674
and this is a link to a 64 bit version on that page:
https://github.com/DavidBailes/audacity/suites/15101755792/artifacts/862916456


> Thanks. Frankly, I think it would be better to go back to the behavior of version 3.2.1 and: - to give the "toggleButton" role to the button, - to emit the event which signals the change of state (stateChange) when the "Enter" key is pressed (this > should be possible since today, it is done for the change of name). Best regards. Paul. 

As I've said above, using MSAA there is no role of toggle button available. If there were then there wouldn't be the problem which I've had to work around. The standard behaviour of toggle buttons for screen readers is that when a user tabs to one the screen reader reads whether it is pressed or not pressed (the windows screen readers vary in their wording of this). And if a user presses the button, the new state is read.
Although there is currently no state, I could add a statechange event in addition to the namechange event if that was any help.

David.

**Comment by DavidBailes** [comment]
Hi Paul,
I said above that:

> If the Power button is the focus and in the not pressed state, then when a user presses Enter, Jaws reads "power button pressed pressed" In fact it reads this whole name twice for reasons unknown.

Concerning Jaws reading the whole name twice after a name change event. I've investigated this further, and found that unfortunately with the latest version of Jaws 2023, this occurs in all the places in Audacity where a name change event is created. I don't  know when this started to happen, but it's not present using Jaws 2022. Unfortunately there are a lot of places where name change events are created.

David.

**Comment by paulber19** [comment]
Hi David.
My thought today is that indicating the "pressed" state in the name of 
the button by signaling it with a nameChange event and positioning this 
state by sending a stateChange event causes an incompatibility that is 
difficult to solve.
The objective of my request was to know if the reading was in progress 
or not. For this, the add-on used the "pressed" state of the "Play" and 
"Pause" buttons.
Since it doesn't seem to be possible anymore, but the status line shows 
if the reading is in progress, why not use it?
For example by placing , at the beginning of this line, an indication (a 
character or symbol), independent of the audacity language.
This must be easily achievable and would not affect the existing buttons.
Best regards.
Paul.

**Comment by DavidBailes** [comment]
Hi Paul,
with regards to your suggestion of putting a character or symbol at the start of the status. I wouldn't want to introduce such a character in the status string which is displayed, so it would have to be in one of the strings of the accessibility object. I've had a quick look at the code, and it doesn't look like it would be that easily achievable, given how the code is structured.
My strong preference, at least for now, is to implement my initial suggestion of adding a character to the end of the accessibility name of the toggle button when it is in its pressed state. I can do that quickly and easily, in time for version 3.4.0
In addition it is a more general fix for the toggle buttons, and allows screen reader scripts, either Jaws or NVDA, to determine the state of any of these toggle buttons if there is the need to do so.
David.

**Comment by paulber19** [comment]
Hi David.
Effectively, this is the most general solution for toggle button.
I will adapt my scripts accordingly.
However, care must be taken that the character is not part of the name 
of the button to be translated and that it is added after its translation.
Best regards.
Paul.

**Comment by DavidBailes** [comment]
Hi Paul,
I've submitted a pull request to add a character to the accessibility name when a toggle button is pressed: https://github.com/audacity/audacity/pull/5066
I've used the '\a' character, which is read silently by all the windows screen readers.

David.

**Comment by DavidBailes** [comment]
@LWinterberg . There's a simple PR to fix this, please can this be included in 3.4?
